### ISSUE TITLE
unhelpful message (to stdout/stderr, as reported by cron via email): Encountered vhost ambiguity but unable to ask for user guidance in non-interactive mode

### ISSUE THREAD

**Issue by jsoref** [certbot/certbot#5139]
## My operating system is (include version):
debian 9.1

## I installed Certbot with (certbot-auto, OS package manager, pip, etc):
certbot-auto (certbot version: 0.18.2)

## I ran this command and it produced this output:
(for my reference, this was apricot @ 2017-09-23 01:13:07,943)
`certbot-auto -q renew` (in a crontab)
```
Encountered vhost ambiguity but unable to ask for user guidance in non-interactive mode. Certbot may need vhosts to be explicitly labelled with ServerName or ServerAlias directives.
Falling back to default vhost *:443...
```

## Certbot's behavior differed from what I expected because:
1. It doesn't indicate which domain is failing.
2. It should distinguish between "I found multiple vhosts that could be responsible for this domain", "I couldn't find any vhosts responsible for this domain", and "there used to be a vhost responsible for this domain, but it's missing/no longer covers this domain".
3. if a cron task produces no output, no email is sent. In general, a cron job should only produce output if an error has occurred.

certbot should probably conceptually queue console warnings like this one until it runs into an instance where it actually needed to ask and couldn't, as opposed to saying "well, I might, eventually, maybe, need some information, maybe, so why don't I bother the user needlessly now, because being eager is a really useful activity". In this case, no error occurred, and thus a warning like this is spurious and really frustrating.

###### Logs are stored in `/var/log/letsencrypt` by default. Feel free to redact domains, e-mail and IP addresses as you see fit.
```
2017-09-23 01:13:07,554:INFO:certbot.auth_handler:Performing the following challenges:
2017-09-23 01:13:07,554:INFO:certbot.auth_handler:tls-sni-01 challenge for example.com
2017-09-23 01:13:07,943:WARNING:certbot_apache.display_ops:Encountered vhost ambiguity but unable to ask for user guidance in non-interactive mode. Certbot may need vhosts to be explicitly labelled with ServerName or ServerAlias directives.
2017-09-23 01:13:07,943:WARNING:certbot_apache.tls_sni_01:Falling back to default vhost *:443...
```

## Here is the relevant nginx server block or Apache virtualhost for the domain I am configuring:
There isn't a vhost for this domain at this time... I'm not sure why. Mostly this domain is only used for email, and certbot is responsible for getting the certificate/renewing it so that dovecot/whatever can serve it.

PS. It did manage to renew the certificate, so, mostly this warning is incredibly disturbing, but not remotely helpful. As someone who gets a lot of email from cron/logwatch, this is an email I really could have done without.

**Comment by SwartzCr** [comment]
Seems mildly confusing, but not like the end of the world, especially since certbot managed to get a cert
@bmw  do you think we should move these from warn to debug?

**Comment by jsoref** [comment]
FWIW, my expectations relate to http://www.linfo.org/rule_of_silence.html -- my `-q` really should be enough to enforce that expectation (although the rule of silence would argue that I shouldn't need `-q` at all).

**Comment by bmw** [comment]
> It doesn't indicate which domain is failing.

I agree that it should. Luckily, this was fixed by @SwartzCr in #5105 and will be included in our release next week.

> It should distinguish between "I found multiple vhosts that could be responsible for this domain", "I couldn't find any vhosts responsible for this domain", and "there used to be a vhost responsible for this domain, but it's missing/no longer covers this domain".

We could potentially differentiate between the first two. If there are multiple vhosts with `ServerName`/`ServerAlias` directives matching the domain, we could give an error like the former, while if there are no such vhosts, we could give one like the latter.

I might be missing some nuance here, but how could we detect the last case? How can Certbot know what your config used to look like?

> if a cron task produces no output, no email is sent. In general, a cron job should only produce output if an error has occurred.

I'm not sure if I agree with this as a general statement, but I agree with you that in this case, if the challenge was successful, the email was largely just noise. I think reducing the logging level is fine.

**Comment by jsoref** [comment]
> I might be missing some nuance here, but how could we detect the last case?
> How can Certbot know what your config used to look like?

When certbot updates its /etc/letsencrypt/renewal files, it can mark which file contained the relevant vhost... (obviously that won't work retroactively, but for most users, their vhosts won't yet have disappeared, so at an update, it records it, and when the vhost later disappears, it can explain where it last saw it)

**Comment by bmw** [comment]
Got it. My personal feeling is that Certbot telling you that you used to have a matching vhost isn't worth the additional complexity of tracking this in case the situation arises and telling you that it cannot find a matching vhost for this domain is enough.

ccing @joohoi as he's currently the main person working on our Apache plugin.

**Comment by joohoi** [comment]
The vhost matching isn't relevant at all for the renew or the actual verification operation. What I think we could do here is to proceed in a similar way that we're doing currently if certbot-apache is run without Installer while specifying the domain, eg:
`certbot certonly --apache -d whatever.yourdomain.tld`

This requires a bit of decoupling and method call reordering, but I'll look into this. There's currently a lot on my plate, and this isn't super high priority but will give it attention at some point.

**Comment by joohoi** [comment]
This was fixed in #5105
